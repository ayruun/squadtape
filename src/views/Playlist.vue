<template>
  <div class="playlist-view">
    <TheInfoBox
      v-if="playlist"
      :id="playlistId"
      :name="playlist.name"
      :description="playlist.description"
      :total="playlist.tracks.total"
      :duration="totalLength"
    />

    <audio
      ref="audio"
      :src="activeTrack"
      @ended="activeTrack = null"
    />

    <Grid v-if="playlist">
      <GridTile
        v-for="{ track } in playlist.tracks.items"
        :key="track.id"
        :clickable="!!track.preview_url"
      >
        <img
          :src="track.album.images[1].url"
          :alt="track.album.name"
          @click="track.preview_url && togglePlay(track.preview_url)"
        />

        <div class="track-info">
          <p>Artist: {{ track.artists[0].name }}</p>
          <p>Track: {{ track.name }}</p>
          <p>Duration: {{ msToTime(track.duration_ms) }}</p>

          <button
            v-if="track.preview_url"
            class="controls"
            @click="togglePlay(track.preview_url)"
            @mouseenter="color = 'var(--secondary)'"
            @mouseleave="color = 'var(--primary)'"
          >
            <IconPlay
              v-if="activeTrack !== track.preview_url"
              :color="color"
            />
            <IconPause
              v-else
              :color="color"
            />
          </button>
        </div>
      </GridTile>
    </Grid>

    <TheFooter />
  </div>
</template>

<script>
import TheInfoBox from "../components/TheInfoBox";
import Grid from "../components/Grid";
import GridTile from "../components/GridTile.vue";
import TheFooter from "../components/TheFooter.vue";
import playlists from "../data/playlistData.json";
import IconPlay from "../components/icons/IconPlay";
import IconPause from "../components/icons/IconPause";

export default {
  name: "App",
  components: {
    TheInfoBox,
    Grid,
    GridTile,
    TheFooter,
    IconPlay,
    IconPause
  },
  props: {
    playlistId: {
      type: String,
      required: true
    }
  },
  data() {
    return {
      playlist: null,
      activeTrack: "",
      color: "var(--primary)"
    };
  },
  computed: {
    totalLength() {
      if (!this.playlist) return null;

      let durationMs = this.playlist.tracks.items.reduce(
        (total, el) => total + el.track.duration_ms,
        0
      );

      return this.msToTime(durationMs);
    }
  },
  created() {
    this.playlist = playlists.find(playlist => playlist.id === this.playlistId);
  },
  methods: {
    msToTime(duration) {
      var seconds = parseInt((duration / 1000) % 60),
        minutes = parseInt(duration / (1000 * 60));

      minutes = minutes < 10 ? "0" + minutes : minutes;
      seconds = seconds < 10 ? "0" + seconds : seconds;

      return `${minutes}:${seconds} min`;
    },
    togglePlay(previewUrl) {
      if (this.activeTrack == previewUrl) {
        if (this.$refs.audio.paused) {
          this.$refs.audio.play();
        } else {
          this.$refs.audio.pause();
          this.activeTrack = null;
        }
      } else {
        this.activeTrack = previewUrl;
        this.$nextTick(() => {
          this.$refs.audio.play();
        });
      }
    }
  }
};
</script>

<style>
.playlist-view {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.controls {
  margin-top: 5px;
  height: 50px;
  width: 50px;
  border-radius: 50%;
  border: none;
  line-height: 50px;
  background-color: var(--transparent);
  cursor: pointer;
}

.controls:focus {
  outline: 0;
}

.track-info {
  position: absolute;
  width: 90%;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  color: var(--primary);
  opacity: 0;
}
</style>
