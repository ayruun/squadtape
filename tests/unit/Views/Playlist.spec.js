import { shallowMount } from '@vue/test-utils'
import Playlist from '@/views/Playlist.vue'

describe('Playlist.vue', () => {
  let playlistId = "6id5t7Oao2KXLXYf1TG4MZ";

  const factory = options => {
    return shallowMount(Playlist, {
      propsData: {
        playlistId
      },
      ...options
    });
  };

  describe('Props', () => {
    it('has one prop', () => {
      const wrapper = factory();
      expect(Object.keys(wrapper.vm.$options.props).length).toBe(1);
    })
    test('playlistId', () => {
      const playlistId = factory().vm.$options.props.playlistId;
      expect(playlistId.type).toBe(String);
      expect(playlistId.required).toBe(true);
    })
  })

  describe('methods', () => {
    test('msToTime', () => {
      const msToTime = factory().vm.msToTime;
      expect(msToTime(3000)).toMatch("00:03 min");
      expect(msToTime(300000)).toMatch("05:00 min");
      expect(msToTime(7200000)).toMatch("120:00 min");
    })
    test('togglePlay', () => {
      const wrapper = factory();
      const togglePlay = wrapper.vm.togglePlay;

      togglePlay("previewURL")
      expect(wrapper.vm.activeTrack).toMatch("previewURL");
      wrapper.vm.$nextTick(() => { expect(wrapper.vm.$refs.audio.play).toHaveBeenCalled(); });
      
      togglePlay("previewURL")
      expect(wrapper.vm.activeTrack).toBe(null);
      expect(wrapper.vm.$refs.audio.pause).toHaveBeenCalled();
    })
  })

  describe('computed', () => {
    test('totalLength', () => {
      const wrapper = factory();
      expect(wrapper.vm.totalLength).toBe("37:53 min");
      wrapper.setData({ playlist: null });
      expect(wrapper.vm.totalLength).toBe(null);
    })
  })

  describe('created', () => {
    it('pulls right playlist', () => {
      const wrapper = factory();
      expect(wrapper.vm.playlist.id).toMatch(playlistId);
    })
  })

  describe('snapshots', () => {
    it('renders', () => {
      expect(factory()).toMatchSnapshot();
    })
  })
})
