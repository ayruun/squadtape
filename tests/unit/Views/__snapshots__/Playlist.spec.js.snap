// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlist.vue snapshots renders 1`] = `
<div class="playlist-view"><audio></audio>
  <grid-stub>
    <theinfobox-stub name="Squad Tape 19/01" description="*for best experience* automatisch überblenden auf 4 Sekunden einstellen" total="12" duration="37:53 min" id="6id5t7Oao2KXLXYf1TG4MZ"></theinfobox-stub>
    <gridtile-stub clickable="true"><img src="https://i.scdn.co/image/c89f852ea7ac41d296aa80c706c3cbb692a5af23">
      <div class="track-info">
        <p>Greeny</p>
        <p>Spüre nix</p>
        <p>03:42 min</p> <button class="controls">
          <iconplay-stub color="var(--primary)"></iconplay-stub>
        </button>
      </div>
    </gridtile-stub>
    <gridtile-stub clickable="true"><img src="https://i.scdn.co/image/d8b52d52b2d0eb9070a1bff206bda1bd5834192e">
      <div class="track-info">
        <p>Post Malone</p>
        <p>Wow.</p>
        <p>02:29 min</p> <button class="controls">
          <iconplay-stub color="var(--primary)"></iconplay-stub>
        </button>
      </div>
    </gridtile-stub>
    <gridtile-stub clickable="true"><img src="https://i.scdn.co/image/77eb7c17cafe55027271d1a9e80c178cb719de25">
      <div class="track-info">
        <p>Big K.R.I.T.</p>
        <p>Energy</p>
        <p>03:15 min</p> <button class="controls">
          <iconplay-stub color="var(--primary)"></iconplay-stub>
        </button>
      </div>
    </gridtile-stub>
    <gridtile-stub clickable="true"><img src="https://i.scdn.co/image/d6d87f7d4b55c6f11cb5d618c44df2e9b0c0469c">
      <div class="track-info">
        <p>Edo Saiya</p>
        <p>Beans</p>
        <p>02:11 min</p> <button class="controls">
          <iconplay-stub color="var(--primary)"></iconplay-stub>
        </button>
      </div>
    </gridtile-stub>
    <gridtile-stub clickable="true"><img src="https://i.scdn.co/image/f7a7f1884708de97c458c911e4f53d639e4e5a14">
      <div class="track-info">
        <p>Wale</p>
        <p>Ungrateful &amp; Thankful</p>
        <p>04:05 min</p> <button class="controls">
          <iconplay-stub color="var(--primary)"></iconplay-stub>
        </button>
      </div>
    </gridtile-stub>
    <gridtile-stub clickable="true"><img src="https://i.scdn.co/image/959a1c1fcc69c303d7258b5ff0669ba551f20d77">
      <div class="track-info">
        <p>Kirk Knight</p>
        <p>Not For Nothing</p>
        <p>03:12 min</p> <button class="controls">
          <iconplay-stub color="var(--primary)"></iconplay-stub>
        </button>
      </div>
    </gridtile-stub>
    <gridtile-stub clickable="true"><img src="https://i.scdn.co/image/ba22f6de1682d28641213138ef71d551ea1860bc">
      <div class="track-info">
        <p>Bad Bunny</p>
        <p>200 Mph</p>
        <p>02:50 min</p> <button class="controls">
          <iconplay-stub color="var(--primary)"></iconplay-stub>
        </button>
      </div>
    </gridtile-stub>
    <gridtile-stub clickable="true"><img src="https://i.scdn.co/image/9f93860500bc472539702ff0353eb2737f63f50c">
      <div class="track-info">
        <p>Lil Xan</p>
        <p>Watch Me Fall</p>
        <p>02:47 min</p> <button class="controls">
          <iconplay-stub color="var(--primary)"></iconplay-stub>
        </button>
      </div>
    </gridtile-stub>
    <gridtile-stub clickable="true"><img src="https://i.scdn.co/image/821277105d0c6e509b547f35ee34c55a566308fa">
      <div class="track-info">
        <p>Yungmon</p>
        <p>Danger</p>
        <p>05:06 min</p> <button class="controls">
          <iconplay-stub color="var(--primary)"></iconplay-stub>
        </button>
      </div>
    </gridtile-stub>
    <gridtile-stub clickable="true"><img src="https://i.scdn.co/image/e9157829fe858032a2f7119cf5a263728cc96358">
      <div class="track-info">
        <p>SAM</p>
        <p>Da wo du herkommst</p>
        <p>02:49 min</p> <button class="controls">
          <iconplay-stub color="var(--primary)"></iconplay-stub>
        </button>
      </div>
    </gridtile-stub>
    <gridtile-stub clickable="true"><img src="https://i.scdn.co/image/fdfcd403f07239026f6cb1fc17e508467a8330ac">
      <div class="track-info">
        <p>HDBeenDope</p>
        <p>Way Out</p>
        <p>02:27 min</p> <button class="controls">
          <iconplay-stub color="var(--primary)"></iconplay-stub>
        </button>
      </div>
    </gridtile-stub>
    <gridtile-stub clickable="true"><img src="https://i.scdn.co/image/8285b6e1d0fa9dc8cb769f50be3e47d898b6f4b4">
      <div class="track-info">
        <p>Ramzoid</p>
        <p>OUTSIDE</p>
        <p>02:55 min</p> <button class="controls">
          <iconplay-stub color="var(--primary)"></iconplay-stub>
        </button>
      </div>
    </gridtile-stub>
    <thefooter-stub></thefooter-stub>
  </grid-stub>
</div>
`;
